blackmagic-debug/blackmagic
# Feature: experimental meson buildsystem

## Detailed description

This is a follow up to #1033
After banging my head against the wall that is the build-system once again, I decided to take a new stab at this, the original port was out of date so I ended up redoing it from scratch

This time it is my intent to get this merged

It's an experimental implementation of the build-system for black magic debug firmware in [meson build](https://mesonbuild.com/index.html)
I made an effort to ensure the original Make build-system continues working with no notable changes, this port is meant to work beside it, as an experimental "feature" developers and experienced users can test and work on until it matures

It's very much experimental, this is a fairly dynamic project and meson does not directly provide convenience functions for some of our use cases like it does for others, this means there are multiple ways we could achieving the desired result (without a "recommended" way), this is my second iteration, the original port remain in the original PR I submitted #1033 if you wish to reference it 

The current implementation is fairly functional (I noted experimental above, but that doesn't imply not functional), all probes (with the exception of BMDA 1. and Blackpill-* 2.) are ported and working (compiling, only native was tested), and target support is selectable, most options the Make build system provided were ported over, bootloaders are also available (though some don't compile, but they were not available through make anyway apparently)
Building BMD as a library is not yet implemented, although it is one of the intended use cases for this upgrade

1. BMDA not supported because BMDA it is to be split into its own project compiled against BMD on v2.0, I chose not to spend the effort porting the target/platform

2. Blackpill probes are not ported, the way the Blackpill probes support is added does not align well with the other targets, making it harder to port, there is a way to cleanly port them (that's the point of meson, to make this stuff easier), but it would be destructive to the Make build-system, since this is an experiment and I want to keep the current build-system fully functional, I will hold on that port in a separate PR for now

Libopencm3 was added as a sub-project, because like the original BMP buildsystem, its build system is... rudimentary, there's no easy way to integrate into meson directly, and because there is some code generation porting it to meson (the ideal approach) is not trivial, I started on it but opted for a easier hackier approach, if someone is interesting in doing the port they are welcome to do so (and I would be thankful), the approach I opted for was to use the meson external project module, this is meant to allow integrating external build systems with meson, but they need to align with the `configure -> make` workflow and allow out of tree builds, which this does not, hence the hacks included in the PR to make it work

I tried to be thorough, with comments, consistent style, and generally a clean implementation, "tried" is the keyword here, I'm very much welcome to feedback
I followed the meson style guide (it's not very comprehensive), with the exception of tabs vs spaces, I opted for tabs to maintain consistency with the rest of the code base, and align with personal preference (and dragonmux's too)
I added copyright headers to all build-system files, it was a last minute call, I don't know if it's appropriate, I will leave that to the official maintainers discretion

One note right now the probe/platform configuration is done through project options, there is an alternative approach I like, using the cross-file to declare everything probe/platform related, this gives more purpose to the cross-file and makes it's use clear (less of just something to remember you have to do, with less experienced users not understanding why they have to do it (we can't make meson fetch the correct cross-file automatically, expressly undesired feature from the devs, understandably so if you understand the build-system, but in some cases it would be reeeally useful)), but this means less integration with the meson configuration and options, which provide some useful functionality, and it would cause some code duplication, I'm torn, but for now I am happy with this

### How to test

- Because meson overlays some files on libopencm3, it's easiest to deinit and remove the submodule before
`git submodule deinit --all`
`rm -r lib/libopencm3`

- Create a build directory
`mkdir build`

- Configure the build (cross-file provides the cross compilation toolchain)
`meson setup build/ --cross-file crossfile/arm-none-eabi.ini`

- By default this will configure for the native probe, we can configure other with (note that an invalid probe will print an informative error)
`meson setup build/ --cross-file crossfile/arm-none-eabi.ini -Dprobe=stlink`

- Deploy the ninjas to build (we can also cd to the build dir and setup/build from there ofc)
`ninja -C build`

- To build the bootloader do
`ninja -C build bootloader`

- All binaries can be seen on the top level of the build directory

- To flash the connected probe from the build system with the latest build ;) run
`ninja -C build flash`



## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (`make PROBE_HOST=native`)
* [x] It builds as BMDA (`make PROBE_HOST=hosted`)
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues



replaces #1033
